feat: rpc Filter type

Kuly14 opened the pull request:
Closes #1371, draft pr for the Filter type.

There are some problems with the imports, for example, if we import the Bloom type from `crate::Bloom`, there are missing methods like `Bloom::contains_bloom()`

For now, I just imported them from `ethereum_types` crate, which does solve the problem but introduces a new crate which is not great.

Any input is appreciated. 
Thanks

gakonst commented on crates/primitives/src/filter.rs:
```diff
@@ -0,0 +1,1214 @@
+         for filter in topic_filters.iter() {
+             let mut is_match = false;
+             for maybe_bloom in filter {
+                 is_match = maybe_bloom.as_ref().map(|b| bloom.contains_bloom(b)).unwrap_or(true);
```
I think we should probably add most of the impls from here? https://docs.rs/ethbloom/0.13.0/src/ethbloom/lib.rs.html#174

cc @mattsse

gakonst commented:
@Kuly14 please get the methods from EthBloom required and use them accordingly here, without having us depend on the whole package if possible^^

Kuly14 commented:
> @Kuly14 please get the methods from EthBloom required and use them accordingly here, without having us depend on the whole package if possible^^

@gakonst I'm not sure I follow, should I create a new Bloom type and add the methods from ehtbloom or should I just import the crate?

In the last commit I used the ethbloom crate but it should be possible to just copy some of the methods so we don't have to use the whole crate, although the crate is only like 200 loc.

gakonst commented:
Let's copy only the methods needed from there, no need to import it.

Kuly14 commented:
I added the logic from EthBloom without having to import the crate. I commented out all the unnecessary methods for now. Is it ok like this?

There are some doc-tests failing, will fix after approval.

mattsse commented on crates/primitives/src/filter.rs:
```diff
@@ -0,0 +1,1414 @@
+
+use core::{mem, ops};
+
+use crunchy::unroll;
+
+// 3 according to yellowpaper
+const BLOOM_BITS: u32 = 3;
+const BLOOM_SIZE: usize = 256;
+
+construct_fixed_hash! {
+    pub struct Bloom(BLOOM_SIZE);
+}
+
+/// Returns log2.
+fn log2(x: usize) -> u32 {
```
I'd like to move this to bloom.rs

Kuly14 replied:
Ok no problem.

mattsse reviewed (changes requested):
we already have a bloom type in bloom.rs which is also implemented via `impl_fixed_hash_type`

can we make modifications to that type instead and add the functions there?

filer type lgtm

Kuly14 commented:
I moved the logic to `bloom.rs` and fixed the doc-tests. There are some docs missing for functions from ethers and EthBloom, so for now I added the #![allow(missing_docs)].

mattsse reviewed (approved):
lgtm
